optimizely/config_manager.py

```diff
@@ -0,0 +1,198 @@
+
+    # Return URL if one is provided or use template and SDK key to get it.
+    if url is None:
+      return url_template.format(sdk_key=sdk_key)
+
+    return url
+
+  def get_update_interval(self, update_interval):
+    """ Helper method to determine frequency at which datafile has to be polled and ProjectConfig updated.
+
+    Args:
+      update_interval: Time in seconds optionally sent in by the user.
+    """
+    polling_interval = update_interval or enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    # If polling interval is less than minimum allowed interval then set it to default update interval.
+    if update_interval < enums.ConfigManager.MIN_UPDATE_INTERVAL:
+      self.logger.debug('Invalid update_interval {} provided. Defaulting to {}'.format(
+        update_interval,
+        enums.ConfigManager.DEFAULT_UPDATE_INTERVAL)
+      )
+      polling_interval = enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    return polling_interval
+
+  def set_last_modified(self, response):
+    """ Looks up and sets last modified time based on Last-Modified header in the response.
+
+     Args:
+       response: requests.Response
+     """
+    self.last_modified = response.headers.get(enums.HTTPHeaders.LAST_MODIFIED)
+
+  def set_config(self, response):
+    """ Looks up and sets datafile and config based on response body.
+
+     Args:
+       response: requests.Response
+     """
+    try:
+      response.raise_for_status()
+    except requests_exceptions.HTTPError as err:
+      self.logger.error('Fetching datafile from {} failed. Error: {}'.format(self.datafile_url, str(err)))
+      return
+
+    # Leave datafile and config unchanged if it has not been modified.
+    if response.status_code == http.HTTPStatus.NOT_MODIFIED:
+      self.logger.debug('Not updating config as datafile has not updated since {}.'.format(self.last_modified))
+      return
+
+    self.set_last_modified(response)
+    # TODO(ali): Add validation here to make sure that we do not update datafile and config if not a valid datafile.
+    self._datafile = response.text
+    # TODO(ali): Add notification listener.
+    self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+    self.logger.info('Received new datafile and updated config.')
+
+  def get_config(self):
+    """ Returns instance of ProjectConfig.
+
+    Returns:
+      ProjectConfig.
+    """
+    return self._config
+
+  def fetch_datafile(self):
+    """ Fetch datafile and set ProjectConfig. """
+
+    request_headers = {}
+    if self.last_modified:
+      request_headers[enums.HTTPHeaders.IF_MODIFIED_SINCE] = self.last_modified
+
+    response = requests.get(self.datafile_url, headers=request_headers)
+    self.set_config(response)
+
+  def _run(self):
+    """ Triggered as part of the thread which fetches the datafile and sleeps until next update interval. """
+    while self.is_running:
+      self.fetch_datafile()
```

Comment:
Nit: if we want this to be consistent with other SDKs, it should wait for the previous request to complete before starting the next one.

Reply:
Here `fetch_datafile` will complete before hitting going to the next step.

Reply:
Ah, right. Then I think this implementation has scheduling behavior like JS used to have in this slide:

https://docs.google.com/presentation/d/1d6B3VxZLU06wsg3RPBqZ1vJImvf3BHCnbbtsJUbdKgY/edit#slide=id.p

The timer is not started until the response, but we wanted to standardize on the Java behavior.

Reply:
It could be that making this change is not worth the extra complexity. I am still nit picking.

Reply:
Synced offline about this. For this implementation we are planning to keep it like this.